parlai/agents/rag/modules.py

```diff
@@ -0,0 +1,570 @@
+        """
+        bsz = ys.size(0)
+        seqlen = ys.size(1)
+        inputs = ys.narrow(1, 0, seqlen - 1)
+        if (ys[:, 0] == self.START_IDX).any() and self.generation_model != 'bart':
+            raise AssertionError(
+                "The Beginning of Sentence token is automatically added to the "
+                "label in decode_forced, but you included it in the label. This means "
+                "your model will have a double BOS token, which is probably not what "
+                "you intended."
+            )
+        doc_scores = encoder_states[-1]
+
+        inputs = self.rag_model_interface.get_initial_forced_decoder_input(
+            bsz,
+            inputs,
+            n_docs=doc_scores.size(1) if doc_scores is not None else None,
+            start_idx=self.START_IDX,
+            end_idx=self.END_IDX,
+            input_turns_cnt=encoder_states[2],
+        )
+        latent, _ = self.decoder(inputs, encoder_states)
+        logits = self.output(latent)
+        _, preds = logits.max(dim=-1)
+        return logits, preds  # type: ignore
+
+
+class RagEncoder(TransformerEncoder):
+    """
+    Subclass TransformerEncoder to use additional positions if desired.
+    """
+
+    def __init__(
+        self,
+        opt: Opt,
+        dictionary: DictionaryAgent,
+        embedding: Optional[torch.nn.Embedding] = None,
+        padding_idx: int = 0,
+    ):
+        """
+        RagEncoder initialization.
+
+        The Rag Seq2seq encoder is just a regular encoder
+        """
+        n_init_positions = get_n_positions_from_options(opt) + opt['n_extra_positions']
+        super().__init__(
+            opt=opt,
+            vocabulary_size=len(dictionary),
+            embedding=embedding,
+            padding_idx=padding_idx,
+            reduction_type='none',
+            n_positions=n_init_positions,
+        )
+
+
+class RagDecoder(TransformerDecoder):
+    """
+    RagDecoder is a subclass of TransformerDecoder.
+
+    No further modifications necessary.
+    """
+
+    pass
+
+
+class T5RagModel(RagModel):
+    """
+    T5 For RAG.
+    """
+
+    def __init__(self, opt, dictionary, retriever_shared=None):
+        opt['t5'] = build_t5(opt)
+        if opt['t5_model_parallel']:
+            opt['t5'].parallelize()
+        else:
+            opt['t5'].deparallelize()
+        super().__init__(opt, dictionary, retriever_shared)
+        self.embedding_size = opt['t5'].model_dim
+        self.t5 = opt.pop('t5', None)
+
+    @classmethod
+    def build_encoder(
+        cls,
+        opt: Opt,
+        *args,
+        dictionary: Optional[DictionaryAgent] = None,
+        embedding: Optional[torch.nn.Embedding] = None,
+        encoder_class: Optional[Type] = None,
+        **kwargs,
+    ):
+        return RagModel.build_encoder(
+            opt,
+            encoder=opt['t5'].get_encoder(),
+            encoder_class=ParlaiT5Encoder,
```

Comment:
This is pretty funky. It should be made better by the "swappable subcomponents" change.